ImageLounge/src/DkCore/DkImageStorage.cpp

```diff
@@ -327,41 +340,75 @@ QImage rotateImageFast(const QImage &img, double angle)
-            imgIn = imgIn.convertToFormat(QImage::Format_ARGB32);
-        }
 
 #ifdef WITH_OPENCV
-        QSize size = angle == 180 ? imgIn.size() : imgIn.size().transposed();
-        QImage imgOut = QImage(size, imgIn.format());
-        const cv::Mat matIn = cv::Mat(imgIn.height(), imgIn.width(), CV_8UC4, (uchar *)imgIn.constBits(), imgIn.bytesPerLine());
-        cv::Mat matOut = cv::Mat(imgOut.height(), imgOut.width(), CV_8UC4, imgOut.bits(), imgOut.bytesPerLine());
-
+        cv::RotateFlags rot{};
         if (angle == 180) {
-            cv::rotate(matIn, matOut, cv::ROTATE_180);
+            rot = cv::ROTATE_180;
         } else if (angle == 90) {
-            cv::rotate(matIn, matOut, cv::ROTATE_90_CLOCKWISE);
+            rot = cv::ROTATE_90_CLOCKWISE;
         } else {
-            cv::rotate(matIn, matOut, cv::ROTATE_90_COUNTERCLOCKWISE);
+            rot = cv::ROTATE_90_COUNTERCLOCKWISE;
         }
 
-        return imgOut;
+        // The pixel depth matters, but the color interpretation does not.
+        switch (img.depth()) {
+        case 1:
+            return rotateImageCVMat(img.convertToFormat(QImage::Format_Indexed8), rot, CV_8U);
+        case 8:
+            return rotateImageCVMat(img, rot, CV_8U);
+        case 16:
+            return rotateImageCVMat(img, rot, CV_16U);
+        case 32:
+            return rotateImageCVMat(img, rot, CV_8UC4);
+        case 64:
+            return rotateImageCVMat(img, rot, CV_16UC4);
+        default:
+            return rotateImageCVMat(img.convertToFormat(QImage::Format_ARGB32), rot, CV_8UC4);
+        }
 #else
         if (angle == 180) {
-            QRgb *outData = reinterpret_cast<QRgb *>(imgIn.bits());
-            const int len = imgIn.width() * imgIn.height();
-            std::reverse(outData, outData + len);
-            return imgIn;
+            return img.mirrored(true, true);
         }
 
+        QImage imgIn = img;
+
         if (angle == 90) {
-            return transposeImage(imgIn).mirrored(true, false);
+            switch (img.depth()) {
+            case 1:
+                imgIn = img.convertToFormat(QImage::Format_Indexed8);
+                return transposeImage<uint8_t>(imgIn).mirrored(true, false);
+            case 8:
+                return transposeImage<uint8_t>(imgIn).mirrored(true, false);
+            case 16:
+                return transposeImage<uint16_t>(imgIn).mirrored(true, false);
+            case 32:
+                return transposeImage<uint32_t>(imgIn).mirrored(true, false);
+            case 64:
+                return transposeImage<uint64_t>(imgIn).mirrored(true, false);
+            default:
+                imgIn = img.convertToFormat(QImage::Format_ARGB32);
+                return transposeImage<uint32_t>(imgIn).mirrored(true, false);
+            }
         }
 
         if (angle == 270) {
-            return transposeImage(imgIn.mirrored(true, false));
+            switch (img.depth()) {
+            case 1:
```

Comment:
So this is the same as the 90 case except you mirror it first...why do you need this block, just do the one call to mirror() and use the rotate 90 block.

Reply:
I guess you want to avoid repetition, but that would partly merge the 90- and 270-degree branches, while all the conditionals remain. It makes the code less obvious. 

It currently looks a bit ugly, but it's clear that one block only does 90-degree things, and the other is only related to 270-degree. It's more reading effort for me to parse {conditional mirror} -> {shared transpose} -> {conditional mirror} into {mirror->transpose} and {transpose->mirror}.

Reply:
ok